stats/thresholds.go

```diff
@@ -138,60 +154,99 @@ func NewThresholds(sources []string) (Thresholds, error) {
 }
 
-func (ts *Thresholds) updateVM(sink Sink, t time.Duration) error {
-	ts.Runtime.Set("__sink__", sink)
-	f := sink.Format(t)
-	for k, v := range f {
-		ts.Runtime.Set(k, v)
-	}
-	return nil
-}
-
-func (ts *Thresholds) runAll(t time.Duration) (bool, error) {
-	succ := true
-	for i, th := range ts.Thresholds {
-		b, err := th.run()
+func (ts *Thresholds) runAll(timeSpentInTest time.Duration) (bool, error) {
+	succeeded := true
+	for i, threshold := range ts.Thresholds {
+		b, err := threshold.run(ts.sinked)
 		if err != nil {
 			return false, fmt.Errorf("threshold %d run error: %w", i, err)
 		}
+
 		if !b {
-			succ = false
+			succeeded = false
 
-			if ts.Abort || !th.AbortOnFail {
+			if ts.Abort || !threshold.AbortOnFail {
 				continue
 			}
 
-			ts.Abort = !th.AbortGracePeriod.Valid ||
-				th.AbortGracePeriod.Duration < types.Duration(t)
+			ts.Abort = !threshold.AbortGracePeriod.Valid ||
+				threshold.AbortGracePeriod.Duration < types.Duration(timeSpentInTest)
 		}
 	}
-	return succ, nil
+
+	return succeeded, nil
 }
 
 // Run processes all the thresholds with the provided Sink at the provided time and returns if any
 // of them fails
-func (ts *Thresholds) Run(sink Sink, t time.Duration) (bool, error) {
-	if err := ts.updateVM(sink, t); err != nil {
-		return false, err
+func (ts *Thresholds) Run(sink Sink, duration time.Duration) (bool, error) {
+	// Initialize the sinks store
+	ts.sinked = make(map[string]float64)
+
+	// FIXME: Remove this comment as soon as the stats.Sink does not expose Format anymore,
+	// see: https://github.com/grafana/k6/pull/2251#discussion_r769807203
+	//
+	// As of December 2021, this block reproduces the behavior of the
+	// stats.Sink.Format behavior. As we intend to try to get away from it,
+	// we instead implement the behavior directly here.
+	//
+	// For more details, see https://github.com/grafana/k6/pull/2251#discussion_r769807203
+	switch sinkImpl := sink.(type) {
+	case *CounterSink:
+		ts.sinked["count"] = sinkImpl.Value
+		ts.sinked["rate"] = sinkImpl.Value / (float64(duration) / float64(time.Second))
+	case *GaugeSink:
+		ts.sinked["value"] = sinkImpl.Value
+	case *TrendSink:
+		ts.sinked["min"] = sinkImpl.Min
+		ts.sinked["max"] = sinkImpl.Max
+		ts.sinked["avg"] = sinkImpl.Avg
+		ts.sinked["med"] = sinkImpl.Med
+
+		// Parse the percentile thresholds and insert them in
+		// the sinks mapping.
+		for _, threshold := range ts.Thresholds {
+			if !strings.HasPrefix(threshold.parsed.AggregationMethod, "p(") {
+				continue
+			}
+
+			// Percentile value has already been parsed as a float64, and should be valid.
+			// The error check is only here to protect our future selves.
+			percentile, err := strconv.ParseFloat(trimDelimited("p(", threshold.parsed.AggregationMethod, ")"), 64)
+			if err != nil {
+				return false, fmt.Errorf(
+					"unable to parse percentile value as 64-bit floating point number; "+
+						"reason: %w", err,
+				)
+			}
```

Comment:
The idea was to make a tradeoff. Considering that percentile are a "special-case", as they hold an actual value, and are the only ones doing so, I preferred parsing them as a "validation that the expression makes sense", and extract their value only later "just in time", as they're actually used. That way, I avoided having a dedicated nullable value in the `thresholdExpression`.

I can also envision a version of this where we introduce a `tokenPercentile`, and a `null.Float` (if it exists) in the `thresholdExpression`. We would then in the context of this version iterate over the thresholds, and if the aggregation method is a percentile, then we verify the value is actually set and add it to the sinks. Would you prefer it that way? 😸 

I truly have no preference. I can see on one side, we pre-mash the work, while the other avoids adding abstraction. Let  me know 🐻

Reply:
Hm, looking at it with fresh eyes, yes, you're right, it can be seen either way... I think my first "instinct" is to minimize number of possible errors and parsing which would be with an additional field in struct. It also reduces number of lines :smile: But depending on how thresholds evolve in the future, that field might become obsolete / to be moved, etc.
So yes, I have a slight preference towards the `tokenPercentile` field but not a very strong one, in case you change your mind and would like to keep the current version :slightly_smiling_face:

Reply:
I've modified the implemented design accordingly 👍🏻. The `thresholdExpression` type now has an `AggregationValue` field that's a `null.Float`, and the parser defines a `tokenPercentile` constant too. Let me know if it looks good on your side 😸

Reply:
Yes, I think it "reads" well and easy! Thank you :slightly_smiling_face: